cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRegisterer.java

```diff
@@ -156,7 +164,107 @@ public Pipeline registerPlugins(ETLConfig config, Class errorDatasetType, Datase
       throw new RuntimeException(e);
     }
 
-    return new Pipeline(sourcePluginId, sinksInfo, transformInfos);
+    Map<String, List<String>> connectionsMapString =
+      Maps.transformEntries(connectionsMap, new Maps.EntryTransformer<String, List<ETLStage>, List<String>>() {
+        @Override
+        public List<String> transformEntry(String key, List<ETLStage> stages) {
+          List<String> stageNames = new ArrayList<String>();
+          for (ETLStage stage : stages) {
+            stageNames.add(stage.getName());
+          }
+          return stageNames;
+        }
+      });
+
+    return new Pipeline(sourcePluginId, sinksInfo, transformInfos, connectionsMapString);
+  }
+
+  @VisibleForTesting
+  static Map<String, List<ETLStage>> validateConnections(ETLConfig config) {
+    Map<String, List<ETLStage>> mapSourceToNextStages = new HashMap<>();
+    if (config.getConnections().isEmpty()) {
+      // connections list is empty, default pipeline, source->transforms(serial)-> sinks
+      String previous = config.getSource().getName();
+      for (ETLStage stage : config.getTransforms()) {
+        mapSourceToNextStages.put(previous, ImmutableList.of(stage));
+        previous = stage.getName();
+      }
+      mapSourceToNextStages.put(previous, ImmutableList.copyOf(config.getSinks()));
+    } else {
+      // there are connections provided perform validation
+      // 1) basic validation, # of connections >= (source + #transform + sink - 1)
```

Comment:
this is not right. Consider the pipeline:
`
              --- transform1 ---
              |                         |
source --                          --- sink
              |                         |
              --- transform2 ---
`